irv/irv.py

```diff
@@ -296,20 +293,21 @@ def count_vals(self, tallies: collections.Counter, rund: int = -1) -> collection
         # this might have been more elegant by changing self.ballots so the first
         # column is always such a choice, but I prefer not having to change ballots
         # as this opens up to very subtle errors
-        for i in range(self.ballots.shape[0]):
-            tocount = 0
-            while not self.is_exhausted(i, tocount) and not (self.ballots[i, tocount] in active_candidates):
-                tocount += 1
-            if not self.is_exhausted(i, tocount):
-                new_tallies[self.ballots[i, tocount]] += 1
+        for i in range(len(self.votes)):
+            to_count = 0
+            while not self.is_exhausted(i,to_count) and not (self.votes[i][to_count] in active_candidates):
+                to_count += 1
+            if not self.is_exhausted(i, to_count):
+                new_tallies[self.votes[i][to_count]] += 1
 
         self._logger.info(f"Round {rund}: New tallies are {new_tallies}")
         return new_tallies
 
     def is_exhausted(self, ballot: int, ranking: int) -> bool:
-        if ranking > self.ballots.shape[1]-1 or type(self.ballots[ballot, ranking]) == float:
-            return True
-        return False
+        for single_ballot in self.votes:
```

Comment:
This implementation is incorrect. This algorithm is kind of complex and we can talk about it on Sunday, but the gist is:

- For each IRV round:
  - For each ballot:
      - Get their top candidate
      - If their top candidate is not in the election anymore (`self.votes[i][to_count] in active_candidates`) get the next one. Make that the new top candidate.  Otherwise, add 1 to the vote count for the candidate and go to the next ballot.
      - Go to the next ballot if the voter has no more active candidates (`is_exhausted`)


You have two routes forward:
1. Fix the is_exhausted function
2. Restructure the algorithm that uses it to not need it. 

1 might be faster, but 2 will result in better code. is_exhausted was made because Nathan was using numpy arrays, which had to be rectangular. You don’t have that constraint anymore.